examples/build_dataset/run_build_dataset_with_metadata.py

```diff
@@ -0,0 +1,54 @@
+import dataclasses
+import json
+import sys
+import typing
+
+import hydra
+from hydra.core.config_store import ConfigStore
+from omegaconf import OmegaConf
+from transformers import AutoTokenizer
+
+from bsmetadata.experiments.with_metadata import build_dataset
+from bsmetadata.input_pipeline import DataConfig
+
+
+cs = ConfigStore.instance()
+cs.store(name="config", node=DataConfig)
```

Comment:
I may have missed some steps or run into incompatible dependencies:

```python
> python run_build_dataset_with_metadata.py 
Traceback (most recent call last):
  File "run_build_dataset_with_metadata.py", line 16, in <module>
    cs.store(name="config", node=DataConfig)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/hydra/core/config_store.py", line 85, in store
    cfg = OmegaConf.structured(node)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 119, in structured
    return OmegaConf.create(obj, parent, flags)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 175, in create
    flags=flags,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 861, in _create_impl
    format_and_raise(node=None, key=None, value=None, msg=str(e), cause=e)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 741, in format_and_raise
    _raise(ex, cause)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 719, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set end OC_CAUSE=1 for full backtrace
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 838, in _create_impl
    flags=flags,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 113, in __init__
    format_and_raise(node=None, key=key, value=None, cause=ex, msg=str(ex))
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 741, in format_and_raise
    _raise(ex, cause)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 719, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set end OC_CAUSE=1 for full backtrace
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 97, in __init__
    self._set_value(content, flags=flags)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 638, in _set_value
    raise e
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 635, in _set_value
    self._set_value_impl(value, flags)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 665, in _set_value_impl
    data = get_structured_config_data(value, allow_objects=ao)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 445, in get_structured_config_data
    return get_dataclass_data(obj, allow_objects=allow_objects)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 374, in get_dataclass_data
    node=dummy_parent, key=name, value=value, cause=ex, msg=str(ex)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 741, in format_and_raise
    _raise(ex, cause)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 719, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set end OC_CAUSE=1 for full backtrace
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 370, in get_dataclass_data
    parent=dummy_parent,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 1047, in _maybe_wrap
    ref_type=ref_type,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 999, in _node_wrap
    element_type=element_type,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 113, in __init__
    format_and_raise(node=None, key=key, value=None, cause=ex, msg=str(ex))
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 741, in format_and_raise
    _raise(ex, cause)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 719, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set end OC_CAUSE=1 for full backtrace
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 97, in __init__
    self._set_value(content, flags=flags)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 638, in _set_value
    raise e
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 635, in _set_value
    self._set_value_impl(value, flags)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 665, in _set_value_impl
    data = get_structured_config_data(value, allow_objects=ao)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 445, in get_structured_config_data
    return get_dataclass_data(obj, allow_objects=allow_objects)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 374, in get_dataclass_data
    node=dummy_parent, key=name, value=value, cause=ex, msg=str(ex)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 741, in format_and_raise
    _raise(ex, cause)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 719, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set end OC_CAUSE=1 for full backtrace
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 370, in get_dataclass_data
    parent=dummy_parent,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 1047, in _maybe_wrap
    ref_type=ref_type,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 999, in _node_wrap
    element_type=element_type,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 113, in __init__
    format_and_raise(node=None, key=key, value=None, cause=ex, msg=str(ex))
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 741, in format_and_raise
    _raise(ex, cause)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 719, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set end OC_CAUSE=1 for full backtrace
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 97, in __init__
    self._set_value(content, flags=flags)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 638, in _set_value
    raise e
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 635, in _set_value
    self._set_value_impl(value, flags)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 665, in _set_value_impl
    data = get_structured_config_data(value, allow_objects=ao)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 445, in get_structured_config_data
    return get_dataclass_data(obj, allow_objects=allow_objects)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 374, in get_dataclass_data
    node=dummy_parent, key=name, value=value, cause=ex, msg=str(ex)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 741, in format_and_raise
    _raise(ex, cause)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 719, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set end OC_CAUSE=1 for full backtrace
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 370, in get_dataclass_data
    parent=dummy_parent,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 1047, in _maybe_wrap
    ref_type=ref_type,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 999, in _node_wrap
    element_type=element_type,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 113, in __init__
    format_and_raise(node=None, key=key, value=None, cause=ex, msg=str(ex))
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 741, in format_and_raise
    _raise(ex, cause)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 719, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set end OC_CAUSE=1 for full backtrace
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 97, in __init__
    self._set_value(content, flags=flags)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 638, in _set_value
    raise e
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 635, in _set_value
    self._set_value_impl(value, flags)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/dictconfig.py", line 665, in _set_value_impl
    data = get_structured_config_data(value, allow_objects=ao)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 445, in get_structured_config_data
    return get_dataclass_data(obj, allow_objects=allow_objects)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 374, in get_dataclass_data
    node=dummy_parent, key=name, value=value, cause=ex, msg=str(ex)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 741, in format_and_raise
    _raise(ex, cause)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 719, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set end OC_CAUSE=1 for full backtrace
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 370, in get_dataclass_data
    parent=dummy_parent,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 1047, in _maybe_wrap
    ref_type=ref_type,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 988, in _node_wrap
    ref_type=ref_type,
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/listconfig.py", line 81, in __init__
    format_and_raise(node=None, key=key, value=None, cause=ex, msg=str(ex))
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 821, in format_and_raise
    _raise(ex, cause)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/_utils.py", line 719, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set end OC_CAUSE=1 for full backtrace
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/listconfig.py", line 79, in __init__
    self._set_value(value=content, flags=flags)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/listconfig.py", line 563, in _set_value
    raise e
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/listconfig.py", line 560, in _set_value
    self._set_value_impl(value, flags)
  File "/Users/mike/Library/Caches/pypoetry/virtualenvs/bsmetadata-v1y_XuK5-py3.7/lib/python3.7/site-packages/omegaconf/listconfig.py", line 577, in _set_value_impl
    "Non optional ListConfig cannot be constructed from None"
omegaconf.errors.ValidationError: Non optional ListConfig cannot be constructed from None
    full_key: attributes_to_keep
    object_type=None
```

Reply:
The last commit should solves this

Reply:
@SaulLu Works like a charm! Thank you for such a swift response~